### DISCUSSION TITLE
poly info command does not list components under lower level namespaces

### DISCUSSION THREAD

**Discussion by hugobettmach** [DavidVujic/python-polylith#95] [Ideas]
**Is your feature request related to a problem? Please describe.**
Hi @DavidVujic, first of all this is such a great project, this way of organising python code makes so much sense in many situations.

The issue I came across is related to using multiple namespaces, under the top level namespace (or org_name) as mentioned in https://github.com/DavidVujic/python-polylith/issues/45. This is something I think could be very useful for my use case.

But, when creating something like the example below...
```
├── README.md
├── bases
├── components
│   └── org_name
│       ├── namespace_1
│       │   ├── component_1
│       │   │   ├── __init__.py
│       │   │   └── core.py
│       │   ├── component_2
│       │   │   ├── __init__.py
│       │   │   └── core.py
│       │   └── component_3
│       │       ├── __init__.py
│       │       └── core.py
│       └── namespace_2
│           ├── component_1
│           │   ├── __init__.py
│           │   └── core.py
│           └── component_2
│               ├── __init__.py
│               └── core.py
├── development
├── poetry.lock
├── projects
├── pyproject.toml
```
Using `poetry poly create component --name=namespace_1/component_1`... And then running `poetry poly info`, I get:
```
Workspace summary

projects: 0
components: 2
bases: 0
development: 1

  brick                                                                       development
 ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  namespace_1                                                                      -
  namespace_2                                                                      -
```

**Describe the solution you'd like**
Would be nice `poly info` could list all the components detected under "sub namespaces". Same would apply to bases as well. For example, would like a summary something like the one below. Hopefully this makes sense?

```
Workspace summary

projects: 0
components: 5
bases: 0
development: 1

  brick                                                                       development
 ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  namespace_1                                                                      -
      component_1                                                                  -
      component_2                                                                  -
      component_3                                                                  -
  namespace_2                                                                      -
      component_1                                                                  -
      component_2                                                                  -
```

**Additional context**
Following on https://github.com/DavidVujic/python-polylith/issues/45.

**Comment by DavidVujic** [comment]
Thank you, very interesting ideas @hugobettmach! Today the `namespace_1` and `namespace_2` are categorized as components by the poly tool, but I think it makes sense to support this kind of scenario too.

**Comment by DavidVujic** [comment]
This could possibly be a new `theme`. Identifying bricks using a structure like `<top-namespace>/<sub-namespace>/<brick>`.

How are the bricks included in the pyproject.toml files?

**Reply by hugobettmach** [reply]
For now we are just testing the new setup so we follow your recommendation:
`{include = "<top-namespace>/<sub-namespace>/<brick>", from = "../../components"}`

But could also be nice to have some thing like include all bricks from a sub-namespace, e.g.:
`{include = "<top-namespace>/<sub-namespace>/*", from = "../../components"}`

**Reply by DavidVujic** [reply]
Thanks! ~~I think that you can use the `*` wildcard already~~, if I'm not mistaking it will run the app properly (but the Polylith tooling will probably not recognize the usage).

Edit, similar discussion here: https://github.com/DavidVujic/python-polylith/discussions/89#discussioncomment-5618553

**Comment by DavidVujic** [comment]
I'm going to give the theming a try, but I would recommend you to reconsider the sub-namespacing into a flat structure as suggested in the `loose` theme style. All bricks should probably be on the same level if you want to go all-in Polylith.

If you still want the sub-namespacing, with the current tooling (while waiting for a future version to support it), you could do something like this: put all your bricks in the `bases` and `components` folder, without a sub-namespace. Then create a new component, named after your sub-namespace.


__Example__:

create a new "component" (i.e. your sub-namespace):
```bash
poetry poly create component --name my_namespace_1
```

Import the bricks that should belong to this sub-namespace in the `__init__.py` of the "component" my_namespace_1:
```python
#  the __init__.py file
from org_name import component1, component2

__all__ = ["component1", "component2"]
```

When using the bricks in a sub-namespace, you could import it like this:
```python
#  some other module using the bricks in a sub-namespace
from org_name.my_namespace_1 import component1

...
```

The pyproject.toml would need to add all components needed + the new namespace:
```toml
packages = [
    {include = "org_name/component1", from = "components"},
    {include = "org_name/component2", from = "components"},
    {include = "org_name/my_namespace_1", from = "components"},
   ...
]
```

**Reply by yhkee0404** [reply]
I think this approach still lacks the visibility of bricks like `component1` or `component2` inside `my_namespace_1` using `poly info` and `poly sync` commands because others can depend on the namespace only as a single brick without the commands detecting the bricks hidden inside it. I ended up prepending the namespace to the name of each brick, test, or project of the example repository without hiding them in a brick, following this tip: https://github.com/DavidVujic/python-polylith/discussions/218#discussioncomment-11419824

Then I had to modify many `import` statements like this example:

```python
# "../../components/example/kafka" = "example/kafka"
from example import kafka, log
from example.kafka.core import fetch_consumer_config, fetch_default_config
```

```python
# "../../components/yhkee0404/example_kafka" = "yhkee0404/example_kafka"
from yhkee0404 import example_kafka as kafka, example_log as log
from yhkee0404.example_kafka.core import fetch_consumer_config, fetch_default_config
```

You can see the difference [here](https://github.com/yhkee0404/yhkee0404.github.io/commit/d0e1bb54c5c9a50248e0e8583b1eba20672d405b#diff-4d3a65001ad93f7daff9a7079757bbd7abc0c0e4637a00d0d9f7961e72893c08) where 105 files changed and +981 -210 lines changed and I added 24 LICENSE files to vender or embed each individual brick, test, or project of the example repository in my repository. I do not think this is a desirable developer experience and there could be a way to improve it but others might gain some insight from it hopefully.

**Comment by DavidVujic** [comment]
I understand that you might want bricks with the same name, but in different in sub-namespaces. This could be accomplished in the suggestion above, but with an alias:
```python
#  the __init__.py file of namespace_1
from org_name import some_brick_doing_X as component1
from org_name import some_brick_doing_Y as component2

__all__ = ["component1", "component2"]
```

```python
#  the __init__.py file of namespace_2
from org_name import some_other_brick_doing_X as component1
from org_name import some_other_brick_doing_Y as component2

__all__ = ["component1", "component2"]
```

When importing the namespace, it expose the alias:

```python
# import the component via the namespace, that has been "renamed" (alias)
from org_name.namespace_1 import component1

```

**Reply by hugobettmach** [reply]
Thanks @DavidVujic for your suggestions. This may be all we need :)

**Comment by lhttjdr** [comment]
Thank you for your excellent work.

Although using alias is a good workaround, is there a plan to support namespace in future?

**Reply by DavidVujic** [reply]
Hi @lhttjdr! Currently, I'm not focusing on that part. The general recommendation for Polylith is to have a fairly flat structure: a shared top namespace, and one sub-namespace (that itself can have any number of sub-namespaces/folders). The combination of a top namespace and the upper most sub-namespace becomes a brick, top_ns/brick. 

You can have it namespaced with several sub-namespaces, but Polylith will only recognize the upper most one as a brick. From a functionality aspect it shouldn't be any issues with it.

**Comment by hstravis** [comment]
Here's another usage for sub-namespaces.

Our small team maintains 100+ python repos. We have a few different types of repos:

* reusable libraries (released as pip packages)
* microservices (fastapi)
* Dash apps
* data science packages (Docker images run in Argo workflows)

As you can guess, it has gotten difficult to maintain the code base. We want to move everything to a single monorepo. There are many ways to do it but we like Polylith. The only "problem" we see with Polylith is that we will have 100+ components (and counting). We would like to nest the components so people don't need to navigate 100+ subdirectories to do their work. Some people on the team work primarily on data science bricks. Some focus on reusable libraries. Others focus on microservices and Dash apps. Rather than overwhelming everyone with 100 bricks every time they look at the workspace, it would be nice if components could be organized into sub-namespaces, such as:

```text
...
/components
  /libs       # variety of reusable components - from serializers to domain-specific data parsers 
  /flask      # extensions used by flask web APIs (auth and serialization, for example)
  /dash       # components and styling to keep Dash apps DRY and consistent
  /web_apis/  # shared endpoints (i.e. flask routes used by multiple microservices)
  /analytics  # business-specific analysis packages
  /schemas    # schemas used across the entire stack
...
```

Sub-namespaces can be superficial - visual candy that makes it easier to navigate the code base. We already ensure that each component (i.e. repo) uses a unique namespace. We will continue to do that in a monorepo regardless of how the components are organized on disk.

Different people on the team tend to focus on different types of components. Data scientists focus on `/analytics` and `/schemas` while software engineers tend to focus on the other areas. So we'd like to organize components into sub-namespaces to make the code base easier to navigate. For example, data scientists never need to be concerned about the `flask` sub-namespace and software engineers usually don't need to look at the `analytics` sub-namespace.

I think I understand the previous comments: the multi-project poetry plugin will function properly but some features of the polylith plugin won't work because it won't find components in subnamespaces - I ran into that a few days ago. We'd really like to take advantage of the full capabilities of the polylith plugin. It would be helpful to visualize the bricks and sync dependencies.

Any advice or guidance would be appreciated. Thanks!

**Comment by hstravis** [comment]
@DavidVujic It seems to me that polylith can only scale so far without support for sub-namespaces because workspaces with dozens or hundreds of components will get more and more difficult to work with. It would be helpful to see an example. Do you know of any polylith repos with *many* components?

**Comment by DavidVujic** [comment]
Thank you for your feedback, much appreciated!I think I understand the reasoning about sub-namespaces, and that they should be represented as grouped polylith components. The grouping sounds much like domains, or entry points for the artifacts to build. The way Polylith views code is that it should be possible to share code across all domains and artifacts. That’s why the architecture promotes a flat component structure.

Otherwise it might not be that straightforward to share code between the “flask” domain and the “analytics” domain. It’s at least a risk that the code within the domain is hidden within the domain itself.

Along with components, there’s also bases in Polyllith. Maybe the groups in your example would fit as bases, with the code behind it as components?I know of some monorepos with hundreds of components that work really well with Polylith. It’s built for those kind of scenarios.

Here’s some examples of production systems:Production systems — polylith/clj-poly 0.2.19cljdoc.org
(Writing this from my phone, because I’m away from keyboard for a couple of days)

/David

**Comment by DavidVujic** [comment]
Another try, link to Production systems:


https://cljdoc.org/d/polylith/clj-poly/0.2.19/doc/production-systems



(Sorry about ha poorly formatted response)

**Reply by hstravis** [reply]
Thanks @DavidVujic!

**Comment by DavidVujic** [comment]
@hstravis When I look at your examples again, it seems very much like some of those would be Polylith bases (i.e. entry points), and I would suggest that you have a look at that. I’ll gladly help out if you have any questions or more feedback!


The source code you would have behind the bases would very much fit as small reusable components. I don’t think 100+ components will be a problem for the developers. Usually it is the opposite: quickly identifying features to use. 


The link to production systems I posted before includes a couple of huge monorepos with a lot of bricks (bases and components) and a lot of projects (the artifacts to deploy).


(I realized replying from the email client doesn’t seem to handle responses in threads)

**Reply by hstravis** [reply]
> @hstravis When I look at your examples again, it seems very much like some of those would be Polylith bases (i.e. entry points), and I would suggest that you have a look at that. I’ll gladly help out if you have any questions or more feedback! The source code you would have behind the bases would very much fit as small reusable components. I don’t think 100+ components will be a problem for the developers. Usually it is the opposite: quickly identifying features to use. The link to production systems I posted before includes a couple of huge monorepos with a lot of bricks (bases and components) and a lot of projects (the artifacts to deploy). (I realized replying from the email client doesn’t seem to handle responses in threads)

I spent more time re-organizing the code base. I realized the code can be grouped into shared buckets such as utility functions, data schemas, database repositories, web clients, and service layers (biz logic). Each bucket corresponds to a single component. We also need an app-specific component for each app because there are multiple bases per app - need a place for code shared between app-specific bases. Now there are ~10 components instead of ~100. Relatively easy to navigate and it should encourage contributors to follow the established pattern.

This is better but it is a more disruptive change. Almost all of the code must be refactored. But we are just changing namespaces, there should be no functional changes, and we have adequate test coverage. We will probably run into problems where poor design choices were made - it might be difficult to decouple some things that should not be coupled.

**Comment by DavidVujic** [comment]
I would love to hear more about your progress, feedback and decisions along the way. Let’s keep in touch!
